crash in taskq_thread

I have only been able to see this on Linux systems with code built using clang.  (I've not seen this on any other configuration):

```
Thread 18 "inproc_thr" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffef392700 (LWP 34248)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff73e3801 in __GI_abort () at abort.c:79
#2  0x000000000041f5f9 in nni_plat_abort () at ../src/platform/posix/posix_debug.c:22
#3  0x000000000041727f in nni_panic (fmt=0x43d50f "pthread_mutex_lock: %s") at ../src/core/panic.c:66
#4  0x000000000041fa5d in nni_pthread_mutex_lock (m=0x651678) at ../src/platform/posix/posix_thread.c:83
#5  0x000000000041f9da in nni_plat_mtx_lock (mtx=0x651670) at ../src/platform/posix/posix_thread.c:269
#6  0x000000000041d0f5 in nni_mtx_lock (mtx=0x651670) at ../src/core/thread.c:27
#7  0x000000000041c825 in nni_taskq_thread (self=0x64d550) at ../src/core/taskq.c:62
#8  0x000000000041d428 in nni_thr_wrap (arg=0x64d558) at ../src/core/thread.c:94
#9  0x000000000042055d in nni_plat_thr_main (arg=0x64d558) at ../src/platform/posix/posix_thread.c:393
#10 0x00007ffff79b56db in start_thread (arg=0x7fffef392700) at pthread_create.c:463
#11 0x00007ffff74c488f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

After *hours* I hit a similar crash on macOS:

```
panic: pthread_mutex_lock: Invalid argument
This message is indicative of a BUG.
Report this at http://github.com/nanomsg/nanomsg
1   multistress                         0x00000001000118a6 nni_panic + 518
2   multistress                         0x000000010001e33a nni_pthread_mutex_lock + 58
3   multistress                         0x000000010001e2ba nni_plat_mtx_lock + 42
4   multistress                         0x0000000100017515 nni_mtx_lock + 21
5   multistress                         0x0000000100017070 nni_task_exec + 256
6   multistress                         0x000000010000a146 nni_aio_finish_impl + 182
7   multistress                         0x000000010000a18d nni_aio_finish_synch + 45
8   multistress                         0x000000010003ade1 nni_tcp_pipe_recv_cb + 657
9   multistress                         0x0000000100016c35 nni_taskq_thread + 165
10  multistress                         0x0000000100017848 nni_thr_wrap + 184
11  multistress                         0x000000010001edb6 nni_plat_thr_main + 118
12  libsystem_pthread.dylib             0x00007fff79b6f661 _pthread_body + 340
13  libsystem_pthread.dylib             0x00007fff79b6f50d _pthread_body + 0
14  libsystem_pthread.dylib             0x00007fff79b6ebf9 thread_start + 13
Process 47128 stopped
* thread #2, stop reason = signal SIGABRT
    frame #0: 0x00007fff799a7b6e libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff799a7b6e <+10>: jae    0x7fff799a7b78            ; <+20>
    0x7fff799a7b70 <+12>: movq   %rax, %rdi
    0x7fff799a7b73 <+15>: jmp    0x7fff7999eb00            ; cerror_nocancel
    0x7fff799a7b78 <+20>: retq
Target 0: (multistress) stopped.
```

I'm not yet certain that these are the same problem.

--- a/src/core/aio.c
+++ b/src/core/aio.c
@@ -315,7 +315,6 @@ nni_aio_begin(nni_aio *aio)
 	nni_mtx_lock(&nni_aio_lk);
 	// We should not reschedule anything at this point.
 	if (aio->a_stop) {
-		nni_task_unprep(aio->a_task);
 		aio->a_result = NNG_ECANCELED;
 		nni_mtx_unlock(&nni_aio_lk);
 		return (NNG_ECANCELED);

--- a/src/core/taskq.c
+++ b/src/core/taskq.c
@@ -16,10 +16,8 @@ struct nni_task {
 	void *        task_arg;
 	nni_cb        task_cb;
 	nni_taskq *   task_tq;
-	bool          task_sched;
-	bool          task_run;
-	bool          task_done;
-	bool          task_exec;
+	unsigned      task_busy;
+	bool          task_prep;
 	bool          task_fini;
 	nni_mtx       task_mtx;
 	nni_cv        task_cv;
@@ -50,33 +48,26 @@ nni_taskq_thread(void *self)
 	nni_mtx_lock(&tq->tq_mtx);
 	for (;;) {
 		if ((task = nni_list_first(&tq->tq_tasks)) != NULL) {
-			nni_mtx_lock(&task->task_mtx);
-			task->task_run   = true;
-			task->task_sched = false;
-			nni_mtx_unlock(&task->task_mtx);
 			nni_list_remove(&tq->tq_tasks, task);
 			nni_mtx_unlock(&tq->tq_mtx);
 
 			task->task_cb(task->task_arg);
 
 			nni_mtx_lock(&task->task_mtx);
-			if (task->task_sched || task->task_exec) {
-				// task resubmitted itself most likely.
-				// We cannot touch the rest of the flags,
-				// since the called function has taken control.
-				nni_mtx_unlock(&task->task_mtx);
-			} else {
-				task->task_done = true;
-				nni_cv_wake(&task->task_cv);
-
+			task->task_busy--;
+			if (task->task_busy == 0) {
 				if (task->task_fini) {
 					task->task_fini = false;
 					nni_mtx_unlock(&task->task_mtx);
 					nni_task_fini(task);
-				} else {
-					nni_mtx_unlock(&task->task_mtx);
+
+					nni_mtx_lock(&tq->tq_mtx);
+					continue;
 				}
+				nni_cv_wake(&task->task_cv);
 			}
+			nni_mtx_unlock(&task->task_mtx);
+
 			nni_mtx_lock(&tq->tq_mtx);
 			continue;
 		}
@@ -150,6 +141,35 @@ nni_taskq_fini(nni_taskq *tq)
 }
 
 void
+nni_task_exec(nni_task *task)
+{
+	nni_mtx_lock(&task->task_mtx);
+	if (task->task_prep) {
+		task->task_prep = false;
+	} else {
+		task->task_busy++;
+	}
+	nni_mtx_unlock(&task->task_mtx);
+
+	if (task->task_cb != NULL) {
+		task->task_cb(task->task_arg);
+	}
+
+	nni_mtx_lock(&task->task_mtx);
+	task->task_busy--;
+	if (task->task_busy == 0) {
+		if (task->task_fini) {
+			task->task_fini = false;
+			nni_mtx_unlock(&task->task_mtx);
+			nni_task_fini(task);
+			return;
+		}
+		nni_cv_wake(&task->task_cv);
+	}
+	nni_mtx_unlock(&task->task_mtx);
+}
+
+void
 nni_task_dispatch(nni_task *task)
 {
 	nni_taskq *tq = task->task_tq;
@@ -157,92 +177,29 @@ nni_task_dispatch(nni_task *task)
 	// If there is no callback to perform, then do nothing!
 	// The user will be none the wiser.
 	if (task->task_cb == NULL) {
-		nni_mtx_lock(&task->task_mtx);
-		task->task_sched = false;
-		task->task_run   = false;
-		task->task_exec  = false;
-		task->task_done  = true;
-		nni_cv_wake(&task->task_cv);
-		nni_mtx_unlock(&task->task_mtx);
+		nni_task_exec(task);
 		return;
 	}
-	nni_mtx_lock(&tq->tq_mtx);
 	nni_mtx_lock(&task->task_mtx);
-	task->task_sched = true;
-	task->task_run   = false;
-	task->task_exec  = false;
-	task->task_done  = false;
+	if (task->task_prep) {
+		task->task_prep = false;
+	} else {
+		task->task_busy++;
+	}
 	nni_mtx_unlock(&task->task_mtx);
 
+	nni_mtx_lock(&tq->tq_mtx);
 	nni_list_append(&tq->tq_tasks, task);
 	nni_cv_wake1(&tq->tq_sched_cv); // waking just one waiter is adequate
 	nni_mtx_unlock(&tq->tq_mtx);
 }
 
 void
-nni_task_exec(nni_task *task)
-{
-	if (task->task_cb == NULL) {
-		nni_mtx_lock(&task->task_mtx);
-		task->task_sched = false;
-		task->task_run   = false;
-		task->task_exec  = false;
-		task->task_done  = true;
-		nni_cv_wake(&task->task_cv);
-		nni_mtx_unlock(&task->task_mtx);
-		return;
-	}
-	nni_mtx_lock(&task->task_mtx);
-	if (task->task_exec) {
-		// recursive taskq_exec, run it asynchronously
-		nni_mtx_unlock(&task->task_mtx);
-		nni_task_dispatch(task);
-		return;
-	}
-	task->task_exec  = true;
-	task->task_sched = false;
-	task->task_done  = false;
-	task->task_run   = false;
-	nni_mtx_unlock(&task->task_mtx);
-
-	task->task_cb(task->task_arg);
-
-	nni_mtx_lock(&task->task_mtx);
-	task->task_exec = false;
-	if (task->task_sched || task->task_run) {
-		// cb scheduled a task
-		nni_mtx_unlock(&task->task_mtx);
-		return;
-	}
-	task->task_done = true;
-	nni_cv_wake(&task->task_cv);
-	if (task->task_fini) {
-		task->task_fini = false;
-		nni_mtx_unlock(&task->task_mtx);
-		nni_task_fini(task);
-	} else {
-		nni_mtx_unlock(&task->task_mtx);
-	}
-}
-
-void
 nni_task_prep(nni_task *task)
 {
 	nni_mtx_lock(&task->task_mtx);
-	task->task_sched = true;
-	task->task_done  = false;
-	task->task_run   = false;
-	nni_mtx_unlock(&task->task_mtx);
-}
-
-void
-nni_task_unprep(nni_task *task)
-{
-	nni_mtx_lock(&task->task_mtx);
-	task->task_sched = false;
-	task->task_done  = false;
-	task->task_run   = false;
-	nni_cv_wake(&task->task_cv);
+	task->task_busy++;
+	task->task_prep = true;
 	nni_mtx_unlock(&task->task_mtx);
 }
 
@@ -250,8 +207,7 @@ void
 nni_task_wait(nni_task *task)
 {
 	nni_mtx_lock(&task->task_mtx);
-	while ((task->task_sched || task->task_run || task->task_exec) &&
-	    (!task->task_done)) {
+	while (task->task_busy) {
 		nni_cv_wait(&task->task_cv);
 	}
 	nni_mtx_unlock(&task->task_mtx);
@@ -268,15 +224,12 @@ nni_task_init(nni_task **taskp, nni_taskq *tq, nni_cb cb, void *arg)
 	NNI_LIST_NODE_INIT(&task->task_node);
 	nni_mtx_init(&task->task_mtx);
 	nni_cv_init(&task->task_cv, &task->task_mtx);
-	task->task_sched = false;
-	task->task_done  = false;
-	task->task_run   = false;
-	task->task_sched = false;
-	task->task_exec  = false;
-	task->task_cb    = cb;
-	task->task_arg   = arg;
-	task->task_tq    = tq != NULL ? tq : nni_taskq_systq;
-	*taskp           = task;
+	task->task_prep = false;
+	task->task_busy = 0;
+	task->task_cb   = cb;
+	task->task_arg  = arg;
+	task->task_tq   = tq != NULL ? tq : nni_taskq_systq;
+	*taskp          = task;
 	return (0);
 }
 
@@ -285,7 +238,7 @@ nni_task_fini(nni_task *task)
 {
 	NNI_ASSERT(!nni_list_node_active(&task->task_node));
 	nni_mtx_lock(&task->task_mtx);
-	if ((task->task_run || task->task_exec) && (!task->task_done)) {
+	if (task->task_busy) {
 		// destroy later.
 		task->task_fini = true;
 		nni_mtx_unlock(&task->task_mtx);

--- a/src/core/taskq.h
+++ b/src/core/taskq.h
@@ -31,13 +31,12 @@ extern void nni_task_dispatch(nni_task *);
 // doubt, use nni_task_dispatch instead.)
 extern void nni_task_exec(nni_task *);
 
-// nni_task_prep and nni_task_unprep are used by and exclusively for the aio
-// framework.  nni_task_prep marks the task as "scheduled" without actually
+// nni_task_prep is used by and exclusively for the aio framework.
+// nni_task_prep marks the task as "scheduled" without actually
 // dispatching anything to it yet; nni_task_wait will block waiting for the
 // task to complete normally (after a call to nni_task_dispatch or
-// nni_task_exec), or for nni_task_unprep to be called.
+// nni_task_exec).
 extern void nni_task_prep(nni_task *);
-extern void nni_task_unprep(nni_task *);
 
 extern void nni_task_wait(nni_task *);
 extern int  nni_task_init(nni_task **, nni_taskq *, nni_cb, void *);

